src/emx/src/libsyslog/syslog.3
.\" 
.\" Copyright (c) 2000 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer. 
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name "Carnegie Mellon University" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any other legal
.\"    details, please contact  
.\"      Office of Technology Transfer
.\"      Carnegie Mellon University
.\"      5000 Forbes Avenue
.\"      Pittsburgh, PA  15213-3890
.\"      (412) 268-4387, fax: (412) 268-7395
.\"      tech-transfer@andrew.cmu.edu
.\"
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\" 
.\" $Revision: 1.2 $
.\" Copyright (c) 1985 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted provided
.\" that: (1) source distributions retain this entire copyright notice and
.\" comment, and (2) distributions including binaries display the following
.\" acknowledgement:  ``This product includes software developed by the
.\" University of California, Berkeley and its contributors'' in the
.\" documentation or other materials provided with the distribution and in
.\" all advertising materials mentioning features or use of this software.
.\" Neither the name of the University nor the names of its contributors may
.\" be used to endorse or promote products derived from this software without
.\" specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)syslog.3	6.13 (Berkeley) 6/23/90
.\"
.TH SYSLOG 3 "June 23, 1990"
.UC 5
.SH NAME
syslog, openlog, closelog, setlogmask, vsyslog \- control system log
.SH SYNOPSIS
.B "#include <syslog.h>
.PP
.B "openlog(ident, logopt, facility)
.br
.B "char *ident;
.PP
.B "syslog(priority, message, parameters ... )
.br
.B "int priority;
.br
.B "char *message;
.PP
.B "#include <varargs.h>
.br
.B "vsyslog(priority, message, args)
.br
.B "int priority;
.br
.B "char *message;
.br
.B "va_list args;
.PP
.B "closelog()
.PP
.B "setlogmask(maskpri)
.SH DESCRIPTION
.I Syslog
arranges to write
.I message
onto the system log maintained by
.IR syslogd (8).
The message is tagged with
.IR priority .
The message looks like a
.IR printf (3)
string except that
.B %m
is replaced by the current error message (as referenced by
.IR errno ).
A trailing newline is added if needed.
An alternate form, in which the arguments have already been captured
using the variable-length argument facilities of
.IR varargs (3),
is available under the name
.IR vsyslog .
.PP
This message will be read by
.IR syslogd (8)
and written to the system console, log files, or forwarded to
.I syslogd
on another host as appropriate.
.PP
Priorities are encoded as a
.I facility
and a
.IR level .
The facility describes the part of the system
generating the message.
The level is selected from an ordered list:
.TP
LOG_ALERT
A condition that should be corrected immediately,
such as a corrupted system database.
.TP
LOG_CRIT
Critical conditions,
e.g., hard device errors.
.TP
LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.TP
LOG_EMERG
A panic condition.
This is normally broadcast to all users.
.TP
LOG_ERR
Errors.
.TP
LOG_INFO
Informational messages.
.TP
LOG_NOTICE
Conditions that are not error conditions,
but should possibly be handled specially.
.TP
LOG_WARNING
Warning messages.
.PP
If
.I syslog
cannot pass the message to
.IR syslogd ,
it will attempt to write the message to the console (``/dev/console'')
if the LOG_CONS option is set (see below).
.PP
If special processing is needed,
.I openlog
can be called to initialize the log file.
The parameter
.I ident
is a string that is prepended to every message.
.I Logopt
is a bit field indicating logging options.
Current values for
.I logopt
are:
.TP
LOG_CONS
If unable to send the message to
.IR syslogd ,
write it to the console.
.TP
LOG_NDELAY 
Open the connection to
.I syslogd
immediately.
Normally the open is delayed
until the first message is logged.
Useful for programs that need to manage the
order in which file descriptors are allocated.
.TP
LOG_PERROR
Write the message to stderr as well to the system log.
.TP
LOG_PID
Log the process id with each message:
useful for identifying instantiations of daemons.
.PP
The
.I facility
parameter encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.TP
LOG_AUTH
The authorization system:
.IR login (1),
.IR su (1),
.IR getty (8),
etc.
.TP
LOG_CRON
The clock daemon.
.TP
LOG_DAEMON
System daemons, such as
.IR ftpd (8),
.IR routed (8),
etc, that are not provided for explicitly by other facilities.
.TP
LOG_KERN
Messages generated by the kernel.
These cannot be generated by any user processes.
.TP
LOG_LPR
The line printer spooling system:
.IR lpr (1),
.IR lpc (8),
.IR lpd (8),
etc.
.TP
LOG_MAIL
The mail system.
.TP
LOG_NEWS
The network news system.
.TP
LOG_SYSLOG
Messages generated internally by
.IR syslogd (8).
.TP
LOG_USER
Messages generated by random user processes.
This is the default facility identifier if none is specified.
.TP
LOG_UUCP
The uucp system.
.TP
LOG_LOCAL0
Reserved for local use.
Similarly for LOG_LOCAL1 through LOG_LOCAL7.
.PP
.I Closelog
can be used to close the log file.
.PP
.I Setlogmask
sets the log priority mask to
.I maskpri
and returns the previous mask.
Calls to
.I syslog
with a priority not set in
.I maskpri
are rejected.
The mask for an individual priority
.I pri
is calculated by the macro LOG_MASK(\fIpri\fP);
the mask for all priorities up to and including
.I toppri
is given by the macro LOG_UPTO(\fItoppri\fP).
The default allows all priorities to be logged.
.SH EXAMPLES
.nf
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID, LOG_DAEMON);
setlogmask(LOG_UPTO(LOG_ERR));
syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");
.fi
.SH "SEE ALSO"
logger(1),
syslogd(8)


src/emx/src/libsyslog/syslog.c
/*
 * Copyright (c) 1983, 1988, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)syslog.c	8.5 (Berkeley) 4/29/95";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libc/gen/syslog.c,v 1.29 2003/02/10 08:31:28 alfred Exp $");

#include "namespace.h"
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/syslog.h>
#include <sys/uio.h>
#include <sys/un.h>
#include <netdb.h>
#ifdef __EMX__
# include <netinet/in.h>
#endif

#include <errno.h>
#include <fcntl.h>
#ifndef __EMX__
# include <paths.h>
#else
# define _PATH_CONSOLE "/dev/con"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <stdarg.h>
#include "un-namespace.h"

#include "libc_private.h"

static int      LogFile = -1;           /* fd for log */
static int      connected;              /* have done connect */
static int	opened;			/* have done openlog() */
static int      LogStat = 0;            /* status bits, set by openlog() */
static const char *LogTag = NULL;       /* string to tag the entry with */
static int      LogFacility = LOG_USER; /* default facility code */
static int      LogMask = 0xff;         /* mask of priorities to be logged */

static void	disconnectlog(void); /* disconnect from syslogd */
static void	connectlog(void);	/* (re)connect to syslogd */

#ifndef __EMX__
/*
 * Format of the magic cookie passed through the stdio hook
 */
struct bufcookie {
	char	*base;	/* start of buffer */
	int	left;
};

/*
 * stdio write hook for writing to a static string buffer
 * XXX: Maybe one day, dynamically allocate it so that the line length
 *      is `unlimited'.
 */
static
int writehook(cookie, buf, len)
	void	*cookie;	/* really [struct bufcookie *] */
	char	*buf;		/* characters to copy */
	int	len;		/* length to copy */
{
	struct bufcookie *h;	/* private `handle' */

	h = (struct bufcookie *)cookie;
	if (len > h->left) {
		/* clip in case of wraparound */
		len = h->left;
	}
	if (len > 0) {
		(void)memcpy(h->base, buf, len); /* `write' it. */
		h->base += len;
		h->left -= len;
	}
	return 0;
}
#endif

/*
 * syslog, vsyslog --
 *      print message on log file; output is intended for syslogd(8).
 */
void
syslog(int pri, const char *fmt, ...)
{
        va_list ap;

	va_start(ap, fmt);
	vsyslog(pri, fmt, ap);
	va_end(ap);
}

void
vsyslog(pri, fmt, ap)
	int pri;
	const char *fmt;
	va_list ap;
{
        int cnt;
        char ch, *p;
        time_t now;
        int fd, saved_errno;
	char *stdp, tbuf[2048], fmt_cpy[1024], timbuf[26];
#ifndef __EMX__
	FILE *fp, *fmt_fp;
	struct bufcookie tbuf_cookie;
	struct bufcookie fmt_cookie;
#endif

#define INTERNALLOG     LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID
        /* Check for invalid bits. */
        if (pri & ~(LOG_PRIMASK|LOG_FACMASK)) {
                syslog(INTERNALLOG,
                    "syslog: unknown facility/priority: %x", pri);
                pri &= LOG_PRIMASK|LOG_FACMASK;
        }

        /* Check priority against setlogmask values. */
        if (!(LOG_MASK(LOG_PRI(pri)) & LogMask))
                return;

        saved_errno = errno;

        /* Set default facility if none specified. */
        if ((pri & LOG_FACMASK) == 0)
                pri |= LogFacility;

#ifndef __EMX__
	/* Create the primary stdio hook */
	tbuf_cookie.base = tbuf;
	tbuf_cookie.left = sizeof(tbuf);
	fp = fwopen(&tbuf_cookie, writehook);
	if (fp == NULL)
		return;
#endif

        /* Build the message. */
        (void)time(&now);
#ifndef __EMX__
	(void)fprintf(fp, "<%d>", pri);
	(void)fprintf(fp, "%.15s ", ctime_r(&now, timbuf) + 4);
#else
        p = tbuf + sprintf(tbuf, "<%d>", pri);
        p += sprintf(p, "%.15s ", ctime_r(&now, timbuf) + 4);
#endif
	if (LogStat & LOG_PERROR) {
#ifndef __EMX__
		/* Transfer to string buffer */
		(void)fflush(fp);
		stdp = tbuf + (sizeof(tbuf) - tbuf_cookie.left);
#else
                stdp = p;
#endif
	}
        if (LogTag == NULL)
                LogTag = _getprogname();
        if (LogTag != NULL)
#ifndef __EMX__
            (void)fprintf(fp, "%s", LogTag);
#else
            p += sprintf(p, "%s", LogTag);
#endif
        if (LogStat & LOG_PID)
#ifndef __EMX__
            (void)fprintf(fp, "[%d]", getpid());
#else
            p += sprintf(p, "[%d]", getpid());
#endif
        if (LogTag != NULL) {
#ifndef __EMX__
            (void)fprintf(fp, ": ");
#else
                *p++ = ':';
                *p++ = ' ';
#endif
        }

	/* Check to see if we can skip expanding the %m */
	if (strstr(fmt, "%m")) {
#ifndef __EMX__
		/* Create the second stdio hook */
		fmt_cookie.base = fmt_cpy;
		fmt_cookie.left = sizeof(fmt_cpy) - 1;
		fmt_fp = fwopen(&fmt_cookie, writehook);
		if (fmt_fp == NULL) {
			fclose(fp);
			return;
		}
#else
            char *t = fmt_cpy;
#endif

            /*
             * Substitute error message for %m.  Be careful not to
             * molest an escaped percent "%%m".  We want to pass it
             * on untouched as the format is later parsed by vfprintf.
             */
            for ( ; (ch = *fmt); ++fmt) {
                    if (ch == '%' && fmt[1] == 'm') {
                            ++fmt;
#ifndef __EMX__
                            fputs(strerror(saved_errno), fmt_fp);
#else
                            t += sprintf(t, "%s", strerror(saved_errno));
#endif
                    } else if (ch == '%' && fmt[1] == '%') {
                            ++fmt;
#ifndef __EMX__
                            fputc(ch, fmt_fp);
                            fputc(ch, fmt_fp);
#else
                            *t++ = ch;
                            *t++ = ch;
#endif
                    } else {
#ifndef __EMX__
                        fputc(ch, fmt_fp);
#else
                            *t++ = ch;
#endif
                    }
            }

            /* Null terminate if room */
#ifndef __EMX__
            /* Null terminate if room */
            fputc(0, fmt_fp);
            fclose(fmt_fp);
#else
            *t = '\0';
#endif
            /* Guarantee null termination */
            fmt_cpy[sizeof(fmt_cpy) - 1] = '\0';

            fmt = fmt_cpy;
        }

#ifndef __EMX__
	(void)vfprintf(fp, fmt, ap);
	(void)fclose(fp);

	cnt = sizeof(tbuf) - tbuf_cookie.left;
#else
        cnt = p - tbuf;
        cnt += vsnprintf(p, sizeof(tbuf) - cnt - 1, fmt, ap);
#endif

        /* Output to stderr if requested. */
        if (LogStat & LOG_PERROR) {
                struct iovec iov[2];
		struct iovec *v = iov;

                v->iov_base = stdp;
                v->iov_len = cnt - (stdp - tbuf);
                ++v;
                v->iov_base = "\n";
                v->iov_len = 1;
		(void)_writev(STDERR_FILENO, iov, 2);
        }

	/* Get connected, output the message to the local logger. */
	if (!opened)
		openlog(LogTag, LogStat | LOG_NDELAY, 0);
	connectlog();
	if (send(LogFile, tbuf, cnt, 0) >= 0)
		return;

	/*
	 * If the send() failed, the odds are syslogd was restarted.
	 * Make one (only) attempt to reconnect to /dev/log.
	 */
	disconnectlog();
	connectlog();
	if (send(LogFile, tbuf, cnt, 0) >= 0)
		return;

        /*
	 * Output the message to the console; try not to block
	 * as a blocking console should not stop other processes.
	 * Make sure the error reported is the one from the syslogd failure.
         */
        if (LogStat & LOG_CONS &&
#ifndef __EMX__
	    (fd = _open(_PATH_CONSOLE, O_WRONLY|O_NONBLOCK, 0)) >= 0) {
#else
	    (fd = _open(_PATH_CONSOLE, O_WRONLY/*|O_NONBLOCK*/, 0)) >= 0) {
#endif

		struct iovec iov[2];
		struct iovec *v = iov;

                p = strchr(tbuf, '>') + 1;
		v->iov_base = p;
		v->iov_len = cnt - (p - tbuf);
		++v;
		v->iov_base = "\r\n";
		v->iov_len = 2;
		(void)_writev(fd, iov, 2);
		(void)_close(fd);
        }
}
static void
disconnectlog()
{
	/*
	 * If the user closed the FD and opened another in the same slot,
	 * that's their problem.  They should close it before calling on
	 * system services.
	 */
	if (LogFile != -1) {
		_close(LogFile);
		LogFile = -1;
	}
	connected = 0;			/* retry connect */
}

static void
connectlog()
{
#ifndef __EMX__
	struct sockaddr_un SyslogAddr;	/* AF_UNIX address of local logger */
#else
	struct sockaddr_in SyslogAddr;	/* AF_INET address of local logger */
#endif

	if (LogFile == -1) {
#ifndef __EMX__
		if ((LogFile = _socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
#else
		if ((LogFile = _socket(AF_INET, SOCK_DGRAM, 0)) == -1)
#endif
			return;
		(void)_fcntl(LogFile, F_SETFD, 1);
	}
	if (LogFile != -1 && !connected) {
#ifndef __EMX__
		SyslogAddr.sun_len = sizeof(SyslogAddr);
		SyslogAddr.sun_family = AF_UNIX;
		(void)strncpy(SyslogAddr.sun_path, _PATH_LOG,
		    sizeof SyslogAddr.sun_path);
#else
                memset(&SyslogAddr, 0, sizeof(SyslogAddr));
#ifndef TCPV40HDRS
                SyslogAddr.sin_len = sizeof(SyslogAddr);
#endif
                SyslogAddr.sin_family = AF_INET;
                SyslogAddr.sin_port = htons(514);
                SyslogAddr.sin_addr.s_addr = INADDR_ANY;
#endif
		connected = _connect(LogFile, (struct sockaddr *)&SyslogAddr,
			sizeof(SyslogAddr)) != -1;

#ifndef __EMX__
		if (!connected) {
			/*
			 * Try the old "/dev/log" path, for backward
			 * compatibility.
			 */
			(void)strncpy(SyslogAddr.sun_path, _PATH_OLDLOG,
			    sizeof SyslogAddr.sun_path);
			connected = _connect(LogFile,
				(struct sockaddr *)&SyslogAddr,
				sizeof(SyslogAddr)) != -1;
		}
#endif

		if (!connected) {
			(void)_close(LogFile);
			LogFile = -1;
		}
	}
}

void
openlog(ident, logstat, logfac)
        const char *ident;
        int logstat, logfac;
{
    if (ident != NULL)
            LogTag = ident;
    LogStat = logstat;
    if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
            LogFacility = logfac;

    if (LogStat & LOG_NDELAY)	/* open immediately */
            connectlog();

    opened = 1;	/* ident and facility has been set */
}

void
closelog()
{
	(void)_close(LogFile);
        LogFile = -1;
        LogTag = NULL;
        connected = 0;
}

/* setlogmask -- set the log mask level */
int
setlogmask(pmask)
        int pmask;
{
        int omask;

        omask = LogMask;
        if (pmask != 0)
                LogMask = pmask;
        return (omask);
}
